### PR TITLE
Make barrier sync against current stream instead of whole device

### PR DESCRIPTION
**Pull request by kwen2501**
The document says:

>```torch.distributed.barrier(group=None, async_op=False, device_ids=None)```
>
>Synchronizes all processes.
>This collective blocks processes until the whole group enters this function...

Both `cudaStreamSynchronize` and `cudaDeviceSynchronize` would sync the host-side process to the completion of the barrier kernel in the GPU, hence achieving the above functionality.

This PR chooses a smaller set `cudaStreamSynchronize` instead of `cudaDeviceSynchronize`, for three reasons:
1. lower chance of hang;
2. `CurrentCUDAStream` is usually the context for next operation in Python script;
3. better barrier performance.

### CODE REVIEW

**Comment by wconstab** [comment]
is there a way to get more confidence about this change? or could we offer perhaps an env to restore the old behavior if someone hits a bug?  idk..

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 1]
```diff
@@ -537,10 +537,20 @@ void ProcessGroupNCCL::WorkNCCL::synchronizeInternal(
   // Device synchronize only after we've completed timeout checks.
   if (!barrierTensors_.empty()) {
     // If we use the work to do barrier, we should block here
-    at::cuda::OptionalCUDAGuard gpuGuard;
     for (auto& device : devices_) {
-      gpuGuard.set_index(device.index());
-      AT_CUDA_CHECK(cudaDeviceSynchronize());
+      // `dist.barrier()` only requires all CPU processes to enter this
```
Comment (kumpera) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:541]: We should take this chance to update the documentation of dist::barrier to cover NCCL and explain the expected behavior around CPU and accelerators.

**Review by kumpera — APPROVED** [review]
It would be great if we update the documentation as well to properly explain this behavior.

Additionally, we should note this potentially breaking change in the release notes.

**Review by kumpera — APPROVED** [review]
LGTM.

We should make sure this makes into the release notes for the next PT.

**Pull request closed without merging**